### DISCUSSION TITLE
Index creation in Milvus

### DISCUSSION THREAD

**Discussion by Mahanchello** [milvus-io/milvus#22018] [Q&A and General discussion] [answered]
Hey, I actually have similar case to [this](https://github.com/milvus-io/milvus/discussions/21192 ) and have a couple questions: 
1. Are indexes need to be created for each scalar field if it's used in the search? (e.g. hybrid search)
2. Based on this [config](https://milvus.io/docs/configure_rootcoord.md#rootCoordimportTaskExpiration)  is index gets created by default? How to know/define index type? (had cases when I created a collection, loaded 40K entities, but index didn't get created by default)
3. In case of live inserts into Milvus, at which point new index need to be created? every time the data is inserted? (even for one entity?)
4. When new data gets inserted into collection that previously had an index, should an old index be dropped before creating new one? 
5. Do we need to release collection once the search is done? What if they are a lot of search requests, should collection be loaded every time?

**Comment by yhmo** [comment, marked as answer]
1. Milvus allows user to build index for scalar field, but it is not mandatory. Search performance mainly depends on vector field index, not scalar field index.

2. This [config](https://milvus.io/docs/configure_rootcoord.md#rootCoordimportTaskExpiration) is for "import" operation(we call bulk_insert), it has nothing to do with normal insert operation.
    To define index type(python example):
```
        index_param = {
        "index_type": 'IVF_FLAT',
        "params": {"nlist": 1024},
        "metric_type": "L2"}
        collection.create_index(vector_filed_name, index_param)
```
    To know index type(python example):
```
  indexes = collection.indexes
    for index in indexes:
        print(index.to_dict())
```
There are two ways to create index:
(1). create index before insert operations, milvus will automatically build index once the newly inserted vectors fill an entire segment.
(2). create index after insert operations, the create_index() interface will wait until all the current data be successfully indexed.

**Reply by Mahanchello** [reply]
if index created before/after insert operations with index_type = IVF_SQ8, will subsequent indexes be created having the same index type?

**Reply by yhmo** [reply]
If you call create_index() twice: create_index() ---> create_index(), you will get error: "CreateIndex failed: at most one distinct index is allowed per field".

You need to call drop_index() before the second call of create_index():  create_index() ---> drop_index() ---> create_index(). All the data will be re-indexed by the second index.

**Reply by Mahanchello** [reply]
oh, i see now, thank you so much!

**Comment by yhmo** [comment]
3. The create_index() only needs to be called once. And this method must be called before you call collection.load()

**Comment by yhmo** [comment]
4. Milvus manages data by "segments". When you call insert() to insert data, the data is accumulated into a "growing segment", once a "growing segment" reaches a max size(normally 512Mb), it becomes "sealed" and Milvus will create a new "growing segment" to hold new data. Once a "sealed segment" is generated, Milvus will automatically build for it. Each segment has an independent index.

**Reply by Mahanchello** [reply]
Sorry, confused. `Once a "sealed segment" is generated, Milvus will automatically build for it. ` -> Milvus will build an index automatically? Is index type can be defined for automated creation(e.g. IVF_PQ, etc) or it's going to build IVS_FLAT?

**Reply by yhmo** [reply]
Yes, "Milvus will automatically build index for it."

You need to call drop_index() before the second call of create_index(): create_index() ---> drop_index() ---> create_index(). All the data will be re-indexed by the second index.

For example, firstly I want to create IVF_PQ index, the workflow could be:
1. create a collection
2. create_index(IVF_PQ)
3. insert some data, Milvus build index automatically, some segments were successfully indexed
4. I want to use IVF_FLAT now, I must call collection.drop_index() to drop the old index, then I call collection.create_index(IVF_FLAT) to create new index. After I call create_index(IVF_FLAT), all the "sealed segments" are re-indexed by IVF_FLAT.

**Reply by Mahanchello** [reply]
got it, thank you 👍

**Comment by yhmo** [comment]
5. Just call collection.load() once, Milvus will automatically keep the data of the collection into memory, including the newly inserted data, and do search/query at will. No need to call load() every time before search. When you decide to drop the collection, call collection.release() to free the memory.

**Comment by Mahanchello** [comment]
@yhmo thank you so much for the info, really appreciate it!
